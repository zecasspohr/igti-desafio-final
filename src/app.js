import express from 'express'
import winston from 'winston'

import ClienteRouter from './routes/cliente.route.js'
import AutorRouter from './routes/autor.route.js'
import LivroRouter from './routes/livro.route.js'
import VendaRouter from './routes/venda.route.js'
import auth from './middlewares/auth.js'

const { combine, timestamp, label, printf } = winston.format
const myFormat = printf(({ level, message, label, timestamp }) => {
  return `${timestamp} [${label}] ${level} ${message}`
})
const logger = winston.createLogger({
  level: 'silly',
  transports: [
    new (winston.transports.Console)(),
    new (winston.transports.File)({ filename: 'store-api.log' })
  ],
  format: combine(
    label({ label: 'store-api' }),
    timestamp(),
    myFormat
  )
})
global.logger = logger

const app = express()
app.use(express.json())

app.use(auth.autenticar)

app.use('/cliente', ClienteRouter)
app.use('/autor', AutorRouter)
app.use('/livro', LivroRouter)
app.use('/venda', VendaRouter)

app.use((err, req, res, next) => {
  logger.error(`${req.method} ${req.baseUrl} - ${err.message}`)
  res.status(err.status ? err.status : 400).send({ error: err.message })
})

export default app
