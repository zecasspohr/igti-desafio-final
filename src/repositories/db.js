import Sequelize from 'sequelize'
import dotenv from 'dotenv'
dotenv.config()

const sequelize = new Sequelize(process.env.POSTGRES_CONNECTION,
  {
    dialect: 'postgres',
    define: {
      timestamps: false
    }
  })

export default sequelize
