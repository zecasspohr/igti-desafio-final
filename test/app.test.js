import request from 'supertest'
import app from '../src/app.js'
import db from '../src/repositories/db.js'
import { afterAll, expect, describe, beforeAll, test } from '@jest/globals'
import VendaModel from '../src/models/venda.model.js'

describe('CenÃ¡rio 1', () => {
  const autoraCarolina = {
    nome: 'Carolina Milena Almada',
    email: 'ccarolinamilenaalmada@gmail.com',
    telefone: '83996565550'
  }
  const livroApi = {
    nome: 'APIs em Node.js',
    valor: 90,
    estoque: 25,
    autorId: 1
  }
  const clienteGiovana = {
    nome: 'Giovana Betina Barbosa',
    email: 'giovanabetinabarbosa@gmail.com',
    senha: 'ZjRxDsNQt4',
    telefone: '27998835914',
    endereco: 'Rua Tancredo Neves 639, Serra-ES'
  }

  beforeAll(async () => {
    db.options.logging = false
    await db.query('TRUNCATE vendas, clientes, livros, autores RESTART IDENTITY;')
  })
  afterAll(async () => {
    await db.query('TRUNCATE vendas, clientes, livros, autores RESTART IDENTITY;')
    await db.close()
  })

  test('Cria autor e verifica se foi criado corretamente', async () => {
    await request(app)
      .post('/autor')
      .auth('admin', 'desafio-igti-nodejs')
      .send(autoraCarolina)

    const res = await request(app)
      .get('/autor')
      .auth('admin', 'desafio-igti-nodejs')

    const autorCriado = res.body[0]

    expect(autorCriado).toMatchSnapshot()
    expect(autorCriado.nome).toBe(autoraCarolina.nome)
    expect(autorCriado.email).toBe(autoraCarolina.email)

    autoraCarolina.autorId = autorCriado.autorId
  })
  test('Cria livro e verifica se foi criado corretamente', async () => {
    await request(app)
      .post('/livro')
      .auth('admin', 'desafio-igti-nodejs')
      .send(livroApi)

    const res = await request(app)
      .get('/livro')
      .auth('admin', 'desafio-igti-nodejs')
    const livroCriado = res.body[0]

    expect(livroCriado).toMatchSnapshot()
    expect(livroCriado.nome).toBe(livroApi.nome)
    expect(livroCriado.autorId).toBe(livroApi.autorId)

    livroApi.livroId = livroCriado.livroId
  })
  test('Cria cliente e verifica se foi criado corretamente', async () => {
    await request(app)
      .post('/cliente')
      .auth('admin', 'desafio-igti-nodejs')
      .send(clienteGiovana)

    const res = await request(app)
      .get('/cliente')
      .auth('admin', 'desafio-igti-nodejs')

    const clienteCriado = res.body[0]

    expect(clienteCriado).toMatchSnapshot()
    expect(clienteCriado.nome).toBe(clienteGiovana.nome)
    expect(clienteCriado.email).toBe(clienteGiovana.email)

    clienteGiovana.clienteId = clienteCriado.clienteId
  })

  test('Buscar livro usando usuario cadastrado', async () => {
    const res = await request(app)
      .get('/livro')
      .auth(clienteGiovana.email, clienteGiovana.senha)
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

  test('Cadastrar venda com o cliente e livro criados', async () => {
    const venda = {
      valor: 90,
      data: '2000-08-10',
      clienteId: clienteGiovana.clienteId,
      livroId: livroApi.livroId
    }
    await request(app)
      .post('/venda')
      .auth(clienteGiovana.email, clienteGiovana.senha)
      .send(venda)

    const vendaCriada = await VendaModel.findAll({ raw: true })
    expect(vendaCriada.length).toBe(1)
    expect(vendaCriada).toMatchSnapshot()
  })
})
