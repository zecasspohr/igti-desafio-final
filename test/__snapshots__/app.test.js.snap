// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cenário 1 Buscar livro usando usuario cadastrado 1`] = `
Array [
  Object {
    "autorId": 1,
    "estoque": 25,
    "livroId": 1,
    "nome": "APIs em Node.js",
    "valor": "90",
  },
]
`;

exports[`Cenário 1 Cadastrar venda com o cliente e livro criados 1`] = `
Array [
  Object {
    "clienteId": 1,
    "data": "2000-08-10",
    "livroId": 1,
    "valor": "90",
    "vendaId": 1,
  },
]
`;

exports[`Cenário 1 Cria autor e verifica se foi criado corretamente 1`] = `
Object {
  "autorId": 1,
  "email": "ccarolinamilenaalmada@gmail.com",
  "nome": "Carolina Milena Almada",
  "telefone": "83996565550",
}
`;

exports[`Cenário 1 Cria cliente e verifica se foi criado corretamente 1`] = `
Object {
  "clienteId": 1,
  "email": "giovanabetinabarbosa@gmail.com",
  "endereco": "Rua Tancredo Neves 639, Serra-ES",
  "nome": "Giovana Betina Barbosa",
  "senha": "ZjRxDsNQt4",
  "telefone": "27998835914",
}
`;

exports[`Cenário 1 Cria livro e verifica se foi criado corretamente 1`] = `
Object {
  "autorId": 1,
  "estoque": 25,
  "livroId": 1,
  "nome": "APIs em Node.js",
  "valor": "90",
}
`;
